src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHelpers.cs

```diff
@@ -63,6 +90,25 @@ internal static void CopyBytes(Stream origin, Stream destination, long bytesToCo
             ArrayPool<byte>.Shared.Return(buffer);
         }
 
+        // Asynchronously helps copy a specific number of bytes from one stream into another.
+        internal static async ValueTask CopyBytesAsync(Stream origin, Stream destination, long bytesToCopy, CancellationToken cancellationToken)
+        {
+            byte[] buffer = ArrayPool<byte>.Shared.Rent(minimumLength: MaxBufferLength);
+            while (bytesToCopy > 0)
+            {
+                int currentLengthToRead = (int)Math.Min(MaxBufferLength, bytesToCopy);
+                Memory<byte> memory = buffer.AsMemory(0, currentLengthToRead);
+                int bytesRead = await origin.ReadAsync(memory, cancellationToken).ConfigureAwait(false);
+                if (bytesRead != currentLengthToRead)
+                {
+                    throw new EndOfStreamException();
+                }
+                await destination.WriteAsync(memory, cancellationToken).ConfigureAwait(false);
+                bytesToCopy -= currentLengthToRead;
```

Comment:
```suggestion
                if (bytesRead == 0)
                {
                    throw new EndOfStreamException();
                }
                await destination.WriteAsync(memory.Slice(0, bytesRead), cancellationToken).ConfigureAwait(false);
                bytesToCopy -= bytesRead;
```

Reply:
Yet another place to consume @eerhardt 's good ole ReadExactly* APIs.